rewrite testing across clusters using fixtures

riedel opened the pull request:
* removes considerably more lines than added
* rigorously tests accross all scheduler implementations: tests increase from 168 items to 231 items (not all are actually tested because there is no seperate CI for some)
* move logic for paramaterized testing across multiple schedulers to two fixtures (one that runs locally and one that requires the CI environment): this allows adding new schedulers easily
* add additional required disk argument for HTCondor to fixture using a temporary config default
* test security on all clusters and add xfail for security on HTCondor and Slurm CI
* modify the LocalCluster job_script to use a shebang line in order to behave equally to "real" schedulers (instead of "parsing" the jobscript and "emulating" a shell interpreter)

Please not that this conflicts and is not in sync with @lesteve's #353 . But it might be easier now since #521 was "fixed" .

riedel commented:
IMHO this makes testing alot easier to read

riedel commented on dask_jobqueue/tests/test_job.py:
```diff
@@ -1,61 +1,24 @@
 @pytest.mark.asyncio
-async def test_job(job_cls):
+async def test_job(EnvSpecificCluster):
+    job_cls = EnvSpecificCluster.job_cls
```
The other changes are similar: The logic `pytest.mark.env` logic is moved into the EnvSpecificClusterFixture

riedel commented on dask_jobqueue/tests/test_jobqueue_core.py:
```diff
@@ -15,44 +15,22 @@
-def test_command_template():
-    with PBSCluster(cores=2, memory="4GB") as cluster:
+def test_command_template(Cluster):
+    with Cluster(cores=2, memory="4GB") as cluster:
```
I tried to make that test, that run across all clusters are actually tested across all clusters (or in rare cases only `LocalCluster`)

riedel commented on dask_jobqueue/tests/test_jobqueue_core.py:
```diff
@@ -275,25 +229,22 @@ def __init__(self, *args, **kwargs):
-@pytest.mark.parametrize("Cluster", all_clusters)
 def test_default_number_of_worker_processes(Cluster):
-    with create_cluster_func(Cluster, cores=4, memory="4GB") as cluster:
+    with Cluster(cores=4, memory="4GB") as cluster:
```
Because I moved also the special handling of the `HTCondorCluster ` to the fixture even for local testing things have been becoming much more concise

riedel commented on dask_jobqueue/tests/test_jobqueue_core.py:
```diff
@@ -117,25 +91,6 @@ def test_forward_ip():
         assert cluster.scheduler.ip == default_ip
 
 
-@pytest.mark.parametrize("Cluster", [])
```
Actually I did not know what to make out of those. There are a few later that probably should also be addressed. Eg. also by moving to the apropriate cluster specific testing...

riedel reviewed (commented):
I tried to add a few explanations

riedel commented on dask_jobqueue/tests/test_jobqueue_core.py:
```diff
@@ -407,12 +362,12 @@ def test_wrong_parameter_error(Cluster):
 
-def test_security():
+@pytest.mark.xfail_ci({"htcondor": "#535 no shared filesystem in htcondor ci"})
+@pytest.mark.xfail_ci({"slurm": "#535 no shared filesystem in slurm ci"})
+def test_security(EnvSpecificCluster, loop):
```
This was only tested with LocalCluster before, so that it would not fail although a shared file system is really needed for this to work.

guillaumeeb commented on conftest.py:
```diff
@@ -22,15 +35,29 @@ def pytest_configure(config):
     config.addinivalue_line(
         "markers", "env(NAME): only run test if environment NAME matches"
     )
+    config.addinivalue_line("markers", "xfail_ci(NAME): know CI failure")
```
```suggestion
    config.addinivalue_line("markers", "xfail_ci(NAME): known CI failure")
```

guillaumeeb commented on conftest.py:
```diff
@@ -46,3 +73,36 @@ def mock_lsf_version(monkeypatch, request):
+
+
+@pytest.fixture
+def EnvSpecificCluster3(pytestconfig):
```
Not sure about the final `3` here? Not sure about this definition at all... Could you explain?

guillaumeeb commented on dask_jobqueue/core.py:
```diff
@@ -123,7 +123,7 @@ class Job(ProcessInterface, abc.ABC):
 %(job_header)s
 %(env_header)s
-%(worker_command)s
+exec %(worker_command)s
```
I don't get it, and I have to admit I'm reluctant to add this change.

riedel replied:
I will try to remove this part from the PR. It is a direct consequence of changing this to actually use the script instead of "interpreting" it: https://github.com/dask/dask-jobqueue/blob/8264b8cd6fa028f35665a554c5fea4be34a97781/dask_jobqueue/local.py#L53-L59 . My guess is that it could make a second PR.

guillaumeeb commented on dask_jobqueue/jobqueue.yaml:
```diff
@@ -195,6 +195,7 @@ jobqueue:
     death-timeout: 60           # Number of seconds to wait if a worker can not find a scheduler
     local-directory: null       # Location of fast local storage like /scratch or $TMPDIR
     extra: []
+    shebang: "#!/usr/bin/env bash"
```
Why do we need this?

guillaumeeb commented on dask_jobqueue/local.py:
```diff
@@ -38,11 +38,7 @@ def __init__(
-            config_name=config_name,
-            shebang="",
-            **kwargs
+            scheduler=scheduler, name=name, config_name=config_name, **kwargs
```
Does this work with black linting?

guillaumeeb commented on conftest.py:
```diff
@@ -46,3 +73,36 @@ def mock_lsf_version(monkeypatch, request):
+@pytest.fixture(params=list(all_envs.values()))
+def Cluster(request):
+    if request.param == HTCondorCluster:
+        dask.config.set({"jobqueue.htcondor.disk": "1GB"})
```
Why 1GB here and 20MB above?

guillaumeeb commented on conftest.py:
```diff
@@ -46,3 +73,36 @@ def mock_lsf_version(monkeypatch, request):
+
+
+@pytest.fixture(params=list(all_envs.values()))
+def Cluster(request):
```
So Cluster is all envs, but without specific CI?

riedel replied:
Yes. I thus runs as part of the default pytests without -E . I'll add a comment here to document this.

guillaumeeb reviewed (commented):
This looks really neat!!

I just have a few questions and clarification, and a bit reluctance changing the submission script template.

Otherwise this would be a much welcomed change!

riedel commented:
@guillaumeeb , thanks for reviewing and catching some issues, I tried to incorporate them. I removed the changes to local.py and core.py as they are debatable and can go in a seperate PR . Hope no concerns are left. Added some docstrings to the fixtures.

riedel reviewed (commented):
all your comments should be addressed ?!?

guillaumeeb commented:
Thanks once more @riedel !
